astropy/units/decorators.py

```diff
@@ -100,62 +175,42 @@ def wrapper(*func_args, **func_kwargs):
                 else:
-                    target_unit = param.annotation
-
-                # If the target unit is empty, then no unit was specified so we
-                # move past it
-                if target_unit is not funcsigs.Parameter.empty:
-
-                    if isinstance(target_unit, str):
-
-                        try: # unit passed in as a string
-                            target_unit = Unit(target_unit)
-                            str_target_unit = target_unit.to_string()
-                        except ValueError:
-                            # user specified a physical type instead of a unit
-                            try:
-                                physical_type_id = _unit_physical_mapping[target_unit]
-                                str_target_unit = target_unit
-                            except KeyError:
-                                raise ValueError("Invalid unit of physical type '{0}'."
-                                                 .format(target_unit))
-
-                            target_unit = Unit._from_physical_type_id(physical_type_id)
-
-                    else:
-                        str_target_unit = target_unit.to_string()
-
-                    try:
-                        equivalent = arg.unit.is_equivalent(target_unit,
-                                                  equivalencies=self.equivalencies)
-
-                        if not equivalent:
-                            raise UnitsError("Argument '{0}' to function '{1}'"
-                                             " must be in units convertible to"
-                                             " '{2}'.".format(param.name,
-                                                              wrapped_function.__name__,
-                                                              str_target_unit))
-
-                    # Either there is no .unit or no .is_equivalent
-                    except AttributeError:
-                        if hasattr(arg, "unit"):
-                            error_msg = "a 'unit' attribute without an 'is_equivalent' method"
-                        else:
-                            error_msg = "no 'unit' attribute"
-                        raise TypeError("Argument '{0}' to function '{1}' has {2}. "
-                              "You may want to pass in an astropy Quantity instead."
-                                 .format(param.name, wrapped_function.__name__, error_msg))
+                    targets = param.annotation
+
+                # If the targets is empty, then no target units or physical
+                #   types were specified so we can continue to the next arg
+                if targets is funcsigs.Parameter.empty:
+                    continue
+
+                # If the argument value is None, and the default value is None,
+                #   pass through the None even if there is a target unit
+                if arg is None and param.default is None:
```

Comment:
I agree with @Cadair - I think the default should be checked, and the case where no argument is given but a default is set does get checked naturally because it will be passed as `arg` in the later code blocks.